### PR TITLE
Fiber migration and safety checks

### PR DESCRIPTION
**Pull request by smolt**
Provide a way to live with issue ldc-developers/ldc#666.  The purpose of this pull request is to discuss an approach, and I suppose if it seems ok, it really should go upstream first.

Change has two parts:
1. Prevent crash during context switch if migrating a Fiber to another thread.   Fiber keeps track of its current Thread in an instance var which allows switchIn/switchOut to work correctly even if thread local addresses are cached by the compiler.
2. Add a version CheckFiberMigration that should be set for systems with the issue.  It causes an exception to be thrown if Fiber migration is detected.  A programmer who is careful on such systems can call Fiber.allowMigration() to override the check.  CheckFiberMigration should only be set for those systems with the issue.  A new unittest detects thread local caching and suggests that CheckFiberMigration be set.

OS X has the Fiber migration issue for the CPUs I can test so CheckFiberMigration is initially set for these.

### CODE REVIEW

**Comment by dnadlinger** [comment]
Seems like the best way to go about this for now.

**Comment by smolt** [comment]
@klickverbot  - Can we revert this and I can update?  It won't compile.  The branch was created against 0.15.0 and it uses FiberException which went away.

**Comment by smolt** [comment]
Or should I just make a new PR with updates?

**Comment by kinke** [comment]
It's not thaat crucial as long as LDC master doesn't use it yet. ;) I'd suggest just fixing it directly here on the ldc branch.

**Comment by smolt** [comment]
thanks @kinke, I will do that.

**Comment by smolt** [comment]
done

**Comment by kinke** [comment]
Thx. So according to AppVeyor and Travis (see https://github.com/ldc-developers/ldc/pull/1216), both Win64 and Linux (x86_64 at least) are affected by that issue too (nice idea, the auto-detecting unittest!). Shouldn't we just set `version = CheckFiberMigration` by default for LDC and then emit a recommendation to **un**set it if the issue isn't auto-detected (`tls == 1`)?

**Comment by dnadlinger** [comment]
Let me run this by @liranz. They probably know best how much context switching overhead fiber-heavy applications can tolerate.

**Comment by liranz** [comment]
For us switching fibers quickly is very important, as some of our fibers are only active for a fraction of a micro second, some for very few micros before they are switched out. We depend on the switching to be very quick. For some of our operations we count CPU cycles...

As I stated [previously](http://forum.dlang.org/thread/pflkijjjuyyhextxvdnn@forum.dlang.org?page=1) one of the major uses for fibers is for asynchronous programming, where when all fibers are only part of the same thread (The Reactor) no further synchronization code has to be added, as there is no way of a race between two fibers, and no way of two fibers touching the same variable concurrently.
We use the Linux's `cgroup` mechanism to make sure that the thread that runs our fibers (our reactor) stays local to the same core, and we would never ever move fibers to another thread or another physical core.
If you need more than a single core performance, you can run several reactors, each having its own fibers that would never-ever migrate to another reactor. You will leverage an efficient queue to make sure the reactors are load-balanced.

So, for summary -- I think that even the `Thread              m_curThread;` declaration should be in the `CheckFiberMigration` check, and the non checked version remain as clean as can be without the `m_curThread = tobj;` assignment.

When you start moving fibers around between threads you have to start thinking about synchronization, as races will become possible. Once that has happened, my experience shows that keeping a theradpool to handle incoming work could be made as efficient as the fibers based solution that now also has to take care about synchronization issues. A basic thread pool solution will end up being MUCH simpler to maintain, so using fibers is not worthwhile in this situation. (and in both cases you'll have to care about core affinity, hyper threading, etc to get best performance).
So -- if you're moving fibers around, you should probably change your model.
This last paragraph is true for Linux, I don't have experience with Windows/OS X regarding to performant async. servers.

**Comment by dnadlinger** [comment]
Thanks for repeating this in public here. ;)

I was more wondering about how significant the extra load and branch would be (i.e. whether we should do this on Linux too, and/or have a config-time option to disable the check for Weka-type scenarios). Thinking of it, it would be rather non-trivial.

**Comment by smolt** [comment]
Hey, I did a quick, simplistic analysis.  Context switching performance with this change but CheckFiberMigration unset should be as good as before.

I disassembled and counted x86_64 instructions (OS X with -O3 LLVM 3.6.2), before and after, in same happy path for `call()` and `yield()`, getting essentially same number of subroutine calls, branches taken, and moves:

`call()` 73 instr + `yield()` 31 instr = 104 instr before
`call()` 79 instr + `yield()` 24 instr = 103 instr after

I did not count the instructions in calls to `_d_eh_swapContext()` and `fiber_switchContext()` because they would be the same and make this more in the noise.  Use of `Fiber.m_curThread` eliminates one TLS lookup in `yield()` too, which for some platforms can be significant.

**Comment by smolt** [comment]
And then doing same for when CheckFiberMigration is set:

`call()` 85 instr + `yield()` 24 instr = 109 instr `m_allowMigration=1`
`call()` 86 instr + `yield()` 24 instr = 110 instr `m_allowMigration=0` (default), but 2 extra branches taken

File: src/core/thread.d [hunk 1]
```diff
@@ -5018,6 +5089,80 @@ unittest
...
+    else
+    {
+        assert(!Fiber.migrationUnsafe);
+        // If thread local addr not cached (correct behavior), then tls should
+        // still be 1.
+        assert(tls != 2,
+               "Not safe to migrate Fibers between Threads on your system. "
+               "Consider setting version CheckFiberMigration for this system "
+               "in thread.d");
+        // verify un-cached correct case
```
Comment (kinke) [inline src/core/thread.d:5162]: So how do we get master green again? Convert this failing unittest to a printf() warning?

Comment (smolt) [inline src/core/thread.d:5162]: Yeah, just an informational message makes sense instead of assert.

Anybody go for it, I'm occupied for next 5 hours.

Comment (smolt) [inline src/core/thread.d:5162]: ok, done - but it only got one additional build green.  There is more going on here.

**Pull request merged**